### DISCUSSION TITLE
Virtual Devices

### DISCUSSION THREAD

**Discussion by x1101** [netalertx/NetAlertX#208] [Q&A]
Hello - 
    Part of my network is a wifi mesh, with multiple AP's included. I'd like to be able to create a "virtual device" of some kind for that mesh and be able to connect my wireless devices to that (specifically on the "network" tab, and the "Node" section on the device page. 

Because of the nature of wireless and meshes, they connect to whichever point is closest at any given time, and I'd like to be able to show that on the map more accurately.

Is that something that's doable now in the current setup?

**Comment by jokob-sk** [comment]
Hi!

Thanks for the suggestion!

This is currently not supported however a custom plugin can be written by anyone to do just that:

https://github.com/jokob-sk/Pi.Alert/tree/main/front/plugins

I'd rate this a medium on my priorities list, also I have other projects currently I'm working on so can't give an ETA (can range from days to months depending on what I'm in the mood for doing and community support can sway that too :) ).

Thanks,
j

**Comment by jokob-sk** [comment]
I have thought of a workaround for now:

(make sure you have a backup of the DB/config/original CSV etc.)

1) Export your devices via `Maintenance > Backup / Restore > CSV Export`
2) Add your "Virtual/Dumb device" to the end of the CSV file by copying one of the above lines.
3) Make sure you use a unique MAC address for any manually added lines 
4) Import the modified CSV file via `Maintenance > Backup / Restore > CSV Import`

Hope this helps,
j

**Reply by x1101** [reply]
I did this, changing both the MAC and the IP, and it imported successfully

I did have an import attempt that threw an error (it looks like the one already referenced, but it cleared before I could get a screenshot), but when trying to re-create that issue, it was successful. 

I thing the previously mentioned concern about line endings and EOF characters is a good place to start looking if someone tries this and has issues. 


All that being said, I am still going to dig into how a plugin might do this going forward. 

Thanks for the help!

**Reply by dudefoxlive** [reply]
What did you modify yours with? Maybe excel is doing something weird.

**Reply by x1101** [reply]
I used vim, from a command line

**Reply by dudefoxlive** [reply]
Ok let me try with something thats not excel.

**Comment by dudefoxlive** [comment]
I just attempted to perform this task. I was able to successfully export the file but attempting to modify and import the file keeps failing. I have attached an image of the message that I am getting. The only modifications I did was copy the last line and change the name and Mac Address.
![image](https://user-images.githubusercontent.com/50186495/235328214-5ea7daa9-b4bf-4976-887d-afb884bd86d2.png)

**Comment by jokob-sk** [comment]
1) Can you please verify that the backed-up unmodified CSV file is imported correctly?
2) Use a copy of your backed up CSV file and try to copy the line before the last line (in between existing records). This will make sure there are no trailing white spaces at the end of the file that might cause issues.

**Reply by dudefoxlive** [reply]
> 1. Can you please verify that the backed-up unmodified CSV file is imported correctly?
> 2. Use a copy of your backed up CSV file and try to copy the line before the last line (in between existing records). This will make sure there are no trailing white spaces at the end of the file that might cause issues.

Uploading a file that was just exported works as expected. Attempting to upload a modified file results in errors. This time I only modified the mac address of the line that I just added manually.

**Reply by jokob-sk** [reply]
Can you try using a backup of the exported file and only modify an existing device?
The issues might be caused by different line endings, e.g. if you re using Windows the editor you are using might change line endings that then causes issues in the docker container, that is using Linux ([more info](https://stackoverflow.com/questions/1552749/difference-between-cr-lf-lf-and-cr-line-break-types))

**Comment by Data-Monkey** [comment]
Hi I tried a similar thing but ended up creating a dhcpd.leases file with 2 fake devices in it to represent some dumb switches that I have on my network.
I was thinking of creating a plugin for virtual or fake devices.

Would there be any interest in such a thing?
What devices would you want to add?

Switch
Hub
Mesh AP
Docker Network?

**Reply by dudefoxlive** [reply]
Yes very much. Have a few dumb switches in my network.

**Comment by jokob-sk** [comment]
Closing as released
